### PR TITLE
[CARBONDATA-3720] Support alter table scenario for new insert into flow

### PR DESCRIPTION
**Pull request by ajantha-bhat**
### Why is this PR needed?
Currently, the rearrange logic is based on schema ordinal.

For alter table drop and add columns with/without partition, schema ordinal based re arrange may not work as index will be outside the projection size. This logic become complex to handle.

Hence don't use schema ordinal for rerrange, implemented a position map based rearrange.
 
 ### What changes were proposed in this PR?
implemented a position map based rearrange.
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - No

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/404/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2105/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/408/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2109/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/415/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2116/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/434/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2134/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/443/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala [hunk 1]
```diff
@@ -449,67 +443,48 @@ case class CarbonInsertIntoCommand(databaseNameOp: Option[String],
...
+    var createOrderPosition = 0
+    if (partitionColumnNames != null) {
+      // For alter table drop/add column scenarios, partition column may not be in the end.
+      // Need to keep it in the end.
+      createOrderColumns = createOrderColumns.filterNot(col =>
+        partitionColumnNames.contains(col.getColumnSchema.getColumnName)) ++
+                           createOrderColumns.filter(col =>
+                             partitionColumnNames.contains(col.getColumnSchema.getColumnName))
+    }
+    for (col: CarbonColumn <- createOrderColumns) {
```
Comment (jackylk) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala]: use foreach instead

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala [hunk 2]
```diff
@@ -449,67 +443,48 @@ case class CarbonInsertIntoCommand(databaseNameOp: Option[String],
...
       partitionColumnSchema: mutable.Buffer[ColumnSchema]): (Seq[Int], Seq[ColumnSchema]) = {
     var reArrangedIndex: Seq[Int] = Seq()
     var selectedColumnSchema: Seq[ColumnSchema] = Seq()
-    var complexChildCount: Int = 0
     var partitionIndex: Seq[Int] = Seq()
-    val columnSchema = tableInfo.getFactTable.getListOfColumns.asScala
+    val internalOrderColumns: util.ArrayList[CarbonColumn] = new util.ArrayList[CarbonColumn]()
+    internalOrderColumns.addAll(table.getVisibleDimensions)
+    internalOrderColumns.addAll(table.getVisibleMeasures)
+    val columnSchema = internalOrderColumns.asScala.map(col => col.getColumnSchema)
```
Comment (jackylk) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala]: Can you do like following?
```
val columnSchema = (table.getVisibleDimensions.asScala ++ table.getVisibleDimensions.asScala).map(_.getColumnSchema)
```

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2143/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala (hunk 2 above)
Comment (ajantha-bhat) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala]: yes, can be done like this now. previously it was used for something else

Comment (ajantha-bhat) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala]: done

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala (hunk 1 above)
Comment (ajantha-bhat) [reply integration/spark/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertIntoCommand.scala]: done

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/452/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2153/

**Comment by ajantha-bhat** [comment]
@jackylk : Handled the comments. Build passed. PR is ready.

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**